File: pset6/caesar.py
import sys
import cs50


def main():
    """
    Encrypts a message using Caesar's cipher.

    Usage: caesar k
    Input p

    where k is an integer that acts as key for
    the number of shifts/rotations per alphabetical
    character in the message.
    and
    where p is the string or message to be encrypted.
    """

    # ensure valid arguments
    if len(sys.argv) != 2 or not sys.argv[1].isdigit():
        print("Usage: python caesar.py k")
        exit(1)

    # assign only if key is a nonnegative integer
    k = int(sys.argv[1])

    # get message to encrypt
    print("plaintext: ", end="")
    p = cs50.get_string()

    print("ciphertext: ", end="")
    for c in p:
        # rotate only if alphabetical
        if c.isalpha():
            c = rotate(c, k)
        print(c, end="")
    print("")


def rotate(c, k):
    """Returns character c rotated by k positions."""

    first_letter = get_first_letter(c)
    c = ord(c)

    return chr(((c - first_letter + k) % 26) + first_letter)


def get_first_letter(c):
    """
    Returns ASCII of the first letter of the
    alphabet depending on case.

    Returns ASCII of "a" if lowercase; "A" if uppercase.
    """

    if c.islower():
        return ord("a")
    else:
        return ord("A")


if __name__ == "__main__":
    main()

File: pset2/caesar/caesar.c
/**
 * Encrypts a message using Caesar's cipher.
 *
 * Usage: caesar k
 * Input p
 *
 * where k is an integer that acts as key for
 * the number of shifts/rotations per alphabetical
 * character in the message.
 * and
 * where p is the string or message to be encrypted.
 */

#include <stdio.h>
#include <cs50.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

char rotate(char c, int k);

int main(int argc, string argv[])
{
    // ensure proper usage
    if (argc != 2)
    {
        printf("Usage: ./caesar k\n");
        return 1;
    }

    // get key
    int k = atoi(argv[1]);

    // get message to encrypt
    printf("plaintext: ");
    string p = get_string();

    // iterate over each character in message
    printf("ciphertext: ");
    for (int i = 0, n = strlen(p); i < n; i++)
    {
        // rotate alphabetical character by k positions
        if (isalpha(p[i]))
        {
            p[i] = rotate(p[i], k);
        }
        printf("%c", p[i]);
    }
    printf("\n");

    // success
    return 0;
}

/**
 * Rotates character c by k positions.
 */
char rotate(char c, int k)
{
    // first letter of alphabet based on case
    char first_letter = '\0';
    if (islower(c))
    {
        first_letter = 'a';
    }
    else if (isupper(c))
    {
        first_letter = 'A';
    }

    // calculate new character after shift
    return ((c - first_letter + k) % 26) + first_letter;
}